zkml/docs/src/softmax.md

```diff
@@ -0,0 +1,80 @@
+ Additionally it is assumed that all inputs to the $`\exp`$ lookup are negative, this is to give better stability. This can always be achieved by subtracting a suitable constant before the lookup, and then multiplying the output by a suitable constant after.
+
+ The scaling factor of $`2^{24}`$ means that we can treat inputs as $`32`$-bit numbers with $`8`$-integral bits and $`24`$-fractional bits. We then chunk our inputs into $`8`$-bit limbs, the two most significant are passed to tables that compute the exponential while the two least significant are range checked. The two least significant chunks do not have to bexponentiated as they represent values between $`-(2^{-8})`$ and $`-(2^{-24})`$ and the exponential of any number in this range is ~$`1`$.
+
+ The below image demonstrates how this lookup works:
+
+ ![EXP Lookup](./img/EXP-Lookups.png)
+
+ ### Normalisation
+
+ Proving the normalisation is tricky so rather than doing it explicitly for a vector $`z\in\mathbb{F}^{s}`$ we view it as an integer and dequantise entry-by-entry, via the formula $`\gamma \cdot z_{i} = x_{i}`$. This yields a vector $`x\in\mathbb{R}^{s}`$. Using this we compute
+
+ $$ \hat{z} := \frac{\sqrt{s}}{\gamma}\ln\left(\sum_{j=1}^{s}\exp\left(\frac{x_{j}}{\sqrt{s}}\right)\right). $$
+
+ We then compute $`X = z - \hat{z}`$ element-wise resulting in a vector of values that are all negative. We perform the $`\exp`$ lookup described above on $`X`$ to obtain $`Y`$. If the prover acts honestly then we have $`Y = \mathrm{Softmax}(z)`$. By the fact we used the $`\exp`$ lookup we know that $`Y`$ contains the exponentials of the elements of $`X`$. But $`X`$ was constructed in such a way that its entries should sum to $`1`$. We write $`\theta`$ for the scaling factor of $`Y`$, then observe:
+
+ $$ \begin{align*} \exp\left(\frac{\gamma}{\sqrt{s}}\cdot(z_{i} - \hat{z})\right) &= \exp\left(\frac{\gamma}{\sqrt{s}}\cdot\left(z_{i}-\frac{\sqrt{s}}{\gamma}\ln\left(\sum_{j=1}^{s}\exp\left(\frac{x_{j}}{\sqrt{s}}\right)\right)\right)\right) \\ &= \exp\left(\frac{x_{i}}{\sqrt{s}}-\ln\left(\sum_{j=1}^{s}\exp\left(\frac{x_{j}}{\sqrt{s}}\right)\right)\right) \\ &=\frac{\exp\left(\frac{x_{i}}{\sqrt{s}}\right)}{\sum_{j=1}^{s}\exp\left(\frac{x_{j}}{\sqrt{s}}\right)} \\ &=\frac{1}{\theta}Y_{i}\end{align*} $$
+
+ So if $`Y`$ has been computed correctly then its entries should sum to $`\frac{1}{\theta}`$. We choose our scaling factor $`\frac{1}{\theta}`$ to be $`2^{8}`$, hence we just have to check that the entries of $`Y`$ sum to $`256`$ plus or minus some error.
+
+ ### Allowable Error
+ 
+ Using the formula provided by [zkLLM][1] we can find a suitable bound for the error in normalisation. This error occurs due to rounding in quantisation which can mean the final sum does not always equal $`\frac{1}{\theta}`$. To prove we are in an acceptable range during setup an error bound is determined and we range check the difference between the sum of $`Y`$ and $`\frac{1}{\theta}`$ to be within this range.
```

Comment:
I think we should mention here that we use the same error formulas in zkLLM appendix, but already with K, L and M being set to the values we use to limb-decompose the input to softmax. The code is implementing the generic formulas of zkLLM, which makes sense, but adding this information here will allow to match more easily the specs with the actual code